File: Selenium Web Automation/src/main/java/HomePage.java

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriver;

public class HomePage extends Page {

    private final By cartIcon = By.id("shopping_cart_container");
    private final By productCount = By.className("inventory_item");
    private final By buttonAddCartBackpack = By.id("add-to-cart-sauce-labs-backpack");
    private final By secondHeader = By.className("title");
    private final By cartCountBadge = By.className("shopping_cart_badge");
    private final By buttonRemoveCartBackpack = By.id("remove-sauce-labs-backpack");

    public HomePage(WebDriver driver) {
        super(driver);
    }

    public boolean isCartIconDisplayed() {
        return driver.findElement(cartIcon).isDisplayed();
    }

    public int getProductCount() {
        return driver.findElements(productCount).size();
    }

    public boolean isBackpackButtonDisplayed() {
        try {
            return driver.findElement(buttonAddCartBackpack).isDisplayed();
        } catch (NoSuchElementException e) {
            return false;
        }
    }

    public String getHeaderTitle() {
        return driver.findElement(secondHeader).getText();
    }

    public String getCurrentUrl() {
        return driver.getCurrentUrl();
    }

    public String getPageTitle() {
        return driver.getTitle();
    }

    public void clickAddBackpackButton() {
        driver.findElement(buttonAddCartBackpack).click();
    }

    public boolean isRemoveBackpackButtonDisplayed() {
        return driver.findElement(buttonRemoveCartBackpack).isDisplayed();
    }

    public String getCartCountBadge() {
        return driver.findElement(cartCountBadge).getText();
    }
}

File: WDIO Cucumber Web Automation/features/pageobjects/home.page.js
import { $, expect, browser } from '@wdio/globals'
import Page from './page.js';

class HomePage extends Page {
    get cartIcon() {
        return $("#shopping_cart_container");
    }

    get productCount() {
        return $$('.inventory_item');
    }

    get buttonAddCartBackpack() {
        return $('#add-to-cart-sauce-labs-backpack');
    }

    get buttonRemoveCartBackpack() {
        return $('#remove-sauce-labs-backpack');
    }

    get cartCountBadge() {
        return $('.shopping_cart_badge');
    }

    get secondHeader() {
        return $('.title');
    }

    async currentUrl() {
        return await browser.getUrl();
    }

    async pageTitle() {
        return await browser.getTitle();
    }

    async validateOnHomePage() {
        await expect(this.cartIcon).toBeExisting();
        await expect(this.productCount).toBeElementsArrayOfSize(6);
        await expect(this.buttonAddCartBackpack).toBeEnabled();
        await expect(await this.secondHeader).toHaveText('Products');
        await expect(await this.currentUrl()).toBe('https://www.saucedemo.com/inventory.html');
        await expect(await this.pageTitle()).toBe('Swag Labs');
    }

    async addItem() {
        await this.buttonAddCartBackpack.click();
        //validation
        await expect(this.buttonAddCartBackpack).not.toBeExisting();
        await expect(this.buttonRemoveCartBackpack).toBeExisting();
        await expect(await this.cartCountBadge).toHaveText('1');
    }

    async openCartPage() {
        await this.cartIcon.click();
    }

    open() {
        return super.open("inventory.html");
    }
}

export default new HomePage();